main.go

```diff
@@ -80,7 +122,40 @@ func main() {
 		authorized.POST("/jobs/:id", ctrl.UpdateJob)
 	}
 
-	router.Run()
+	server := http.Server{
+		Addr:    config.Port,
+		Handler: router,
+	}
+
+	serverErrors := make(chan error, 1)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		log.Printf("server listening on port %s", server.Addr)
+		serverErrors <- server.ListenAndServe()
+	}()
+
+	shutdown := make(chan os.Signal, 1)
+	signal.Notify(shutdown, os.Interrupt, syscall.SIGTERM)
+
+	select {
+	case <-serverErrors:
+		return fmt.Errorf("received server error: %w", err)
+
+	case sig := <-shutdown:
+		log.Printf("received shutdown signal %q", sig)
+
+		cancel()
```

Comment:
Is this necessary since we `defer cancel()` above?

Reply:
yes, and here's why.

The defer will ensure that the worker at least gets a shutdown signal if the function returns with an error. More importantly, the defer runs when the function is _just about to return_, which is _after_ the `wg.Wait()` below. If we don't `cancel()` the ctx, then the application will wait forever for the worker to shutdown, which never received the signal to shutdown. Additionally, `cancel()` is safe to call more than once.

Reply:
Appreciate the explanation!